proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy

```diff
@@ -0,0 +1,2912 @@
+(*
+ * Copyright 2022, Proofcraft Pty Ltd
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+theory ArchInvariants_AI
+imports InvariantsPre_AI "Lib.Apply_Trace_Cmd"
+begin
+
+(* setup *)
+
+declare opt_mapE[rule del]
+
+\<comment> \<open>---------------------------------------------------------------------------\<close>
+
+section "AARCH64-specific invariant definitions"
+
+qualify AARCH64 (in Arch)
+record iarch_tcb =
+  itcb_vcpu :: "obj_ref option"
+end_qualify
+
+context Arch begin global_naming AARCH64
+
+(* compatibility with other architectures, input only *)
+abbreviation
+  "vs_lookup s \<equiv> \<lambda>level asid vref. vs_lookup_table level asid vref s"
+
+locale_abbrev
+  "atyps_of \<equiv> \<lambda>s. aobjs_of s ||> aa_type"
```

Comment:
Hm, I can see how this happened. The bit from `iarch_tcb` above these lines is hyp-specific (wouldn't be there in RISC-V), and it needs to be outside the `Arch` locale, because it defines a record. Otherwise these two definitions are at the top of the file, and `iarch_tcb` starts afterwards.

We usually want to avoid entering the `Arch` locale too often, but since this is at the very beginning of AInvs, it should still be reasonably fast. I'll have a look if that is correct.

Reply:
Yes, it's still pretty fast. I've moved these two defs further up into the `setup` section and now things look a bit nicer.